File: python/misc/vanilla_ps1/vanilla_ps1.py
from collections import deque

tests = int(input())
for t in range(tests):
    nodes, edges, start, distance = tuple(map(int, input().split()))
    start -= 1
    graph = [list() for node in range(nodes)]
    visited = [False for node in range(nodes)]
    for e in range(edges):
        left, right = tuple(map(int, input().split()))
        left -= 1
        right -= 1
        graph[left].append(right)
        graph[right].append(left)

    queue = deque()
    # connected components
    num_components = 0
    while False in visited:
        queue.appendleft(visited.index(False))
        num_components += 1
        while queue:
            node = queue.pop()
            if visited[node]: continue
            visited[node] = True
            for a_node in graph[node]:
                if not visited[a_node]:
                    queue.appendleft(a_node)
    # bfs traversal
    num_visited = 0
    visited = [False for node in range(nodes)]
    queue.appendleft((start, 0))
    while queue:
        node, level = queue.pop()
        if level > distance: break
        if visited[node]: continue
        visited[node] = True
        num_visited += 1
        for a_node in graph[node]:
            if not visited[a_node]:
                queue.appendleft((a_node, level + 1))
    print(f'{num_components} {num_visited}')

        

File: cpp/misc/vanilla_ps1/vanilla_ps1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <algorithm>
using namespace std;

int main() {
    int tests;
    cin >> tests;
    while(tests--) {
        int nodes, edges, start, distance;
        scanf("%d %d %d %d", &nodes, &edges, &start, &distance);
        start--;
        vector<vector<int>> graph = vector<vector<int>>(nodes, vector<int>());
        for(int i = 0; i < edges; i++) {
           int left, right;
           cin >> left >> right;
           left--; right--;
           graph[left].push_back(right);
           graph[right].push_back(left);
        }
        int components = 0;
        vector<bool> visited = vector<bool>(nodes, false);
        queue<int> q1;
        while(find(begin(visited), end(visited), false) != end(visited)) {
            components++;
            q1.push(find(begin(visited), end(visited), false) - begin(visited));
            while(!q1.empty()) {
                int node = q1.front(); q1.pop();
                visited[node] = true;
                for(const int &adj_node : graph[node]) {
                    if(!visited[adj_node]) q1.push(adj_node);
                }
            }
        }
        visited = vector<bool>(nodes, false);
        queue<pair<int, int>> q2;
        int num_reachable_nodes = 0;
        q2.push(make_pair(start, 0));
        while(!q2.empty()) {
            int node, level; tie(node, level) = q2.front(); q2.pop();
            if(level > distance) break;
            if(visited[node]) continue;
            num_reachable_nodes++;
            visited[node] = true;
            for(const int adj_node : graph[node]) {
                q2.push(make_pair(adj_node, level+1));
            }
        }
        
        printf("%d %d\n", components, num_reachable_nodes);
    }
}
